File: src/main/java/com/nadmm/airports/wx/WxFragmentBase.java
/*
 * FlightIntel for Pilots
 *
 * Copyright 2012-2021 Nadeem Hasan <nhasan@nadmm.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.nadmm.airports.wx;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.AutoCompleteTextView;
import android.widget.LinearLayout;

import androidx.localbroadcastmanager.content.LocalBroadcastManager;

import com.google.android.material.textfield.TextInputLayout;
import com.nadmm.airports.FragmentBase;
import com.nadmm.airports.utils.UiUtils;

public abstract class WxFragmentBase extends FragmentBase {
    private IntentFilter mFilter;
    private BroadcastReceiver mReceiver;

    @Override
    public void onCreate( Bundle savedInstanceState ) {
        super.onCreate( savedInstanceState );
        setHasOptionsMenu( true );
    }

    @Override
    public void onResume() {
        super.onResume();

        LocalBroadcastManager bm = LocalBroadcastManager.getInstance( getActivity() );
        bm.registerReceiver( mReceiver, mFilter );
    }

    @Override
    public void onPause() {
        super.onPause();

        LocalBroadcastManager bm = LocalBroadcastManager.getInstance( getActivity() );
        bm.unregisterReceiver( mReceiver );
    }

    protected void setupBroadcastFilter( String action ) {
        mFilter = new IntentFilter();
        mFilter.addAction( action );

        mReceiver = new BroadcastReceiver() {

            @Override
            public void onReceive( Context context, Intent intent ) {
                String action = intent.getAction();
                if ( action.equals( mFilter.getAction( 0 ) ) ) {
                    handleBroadcast( intent );
                }
            }
        };
    }

    protected View addWxRow( LinearLayout layout, String label, String code ) {
        View row = addProgressRow( layout, label );
        row.setTag( code );
        int background = UiUtils.getSelectableItemBackgroundResource( getActivity() );
        row.setBackgroundResource( background );
        return row;
    }

    protected AutoCompleteTextView getAutoCompleteTextView( TextInputLayout textInputLayout) {
        return (AutoCompleteTextView) textInputLayout.getEditText();
    }

    protected int getSelectedItemPos( TextInputLayout textInputLayout )
    {
        AutoCompleteTextView textView = getAutoCompleteTextView( textInputLayout );
        if ( textView == null ) return -1;

        ArrayAdapter<?> adapter = (ArrayAdapter<?>) textView.getAdapter();
        if ( adapter == null ) return -1;

        String text = textView.getText().toString();
        if ( text != null && !text.isEmpty() ) {
            int count = adapter.getCount();
            for ( int i = 0; i < count; ++i ) {
                Object o = adapter.getItem( i );
                if ( o.toString().equals( text ) ) return i;
            }
        }
        return -1;
    }

    protected void handleBroadcast( Intent intent ) {
    }

    protected abstract String getProduct();

}

File: src/main/java/com/nadmm/airports/wx/WxFragmentBase.kt
/*
 * FlightIntel for Pilots
 *
 * Copyright 2012-2025 Nadeem Hasan <nhasan@nadmm.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.nadmm.airports.wx

import android.os.Bundle
import android.view.View
import android.widget.AutoCompleteTextView
import android.widget.LinearLayout
import androidx.core.os.BundleCompat
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.google.android.material.textfield.TextInputLayout
import com.nadmm.airports.FragmentBase
import com.nadmm.airports.utils.UiUtils
import kotlinx.coroutines.launch

abstract class WxFragmentBase(protected val action: String) : FragmentBase() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        viewLifecycleOwner.lifecycleScope.launch {
            viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
                NoaaService.Events.events.collect { result ->
                    val resultAction = result.getString(NoaaService.ACTION)
                    if (resultAction == action) {
                        processResult(result)
                    }
                }
            }
        }
    }

    protected fun addWxRow(layout: LinearLayout, label: String, code: String?): View {
        val row = addProgressRow(layout, label)
        row.tag = code
        val background = UiUtils.getSelectableItemBackgroundResource(requireActivity())
        row.setBackgroundResource(background)
        return row
    }

    protected fun getAutoCompleteTextView(textInputLayout: TextInputLayout): AutoCompleteTextView? {
        return textInputLayout.getEditText() as AutoCompleteTextView?
    }

    protected fun getSelectedItemText(textInputLayout: TextInputLayout): String {
        val textView = getAutoCompleteTextView(textInputLayout)
        return textView?.getText().toString()
    }

    protected open fun processResult(result: Bundle) {
    }

    protected fun <T> getResultObject(result: Bundle, clazz: Class<T>): T {
        return BundleCompat.getParcelable(result, NoaaService.RESULT, clazz)
            ?: clazz.getDeclaredConstructor().newInstance()
    }

}
